python/phase-2.md

```diff
@@ -0,0 +1,996 @@
+  attributes which are set in the decorator
+* Don't want to init an instance of the class like this:
+
+    ```python
+    workflow.execute_activity(ActivityClass("init-arg").some_activity, "foo")
+    ```
+* Can create the activity class without initializing it like this:
+
+    ```python
+    activities = ActivityClass.__new__(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
+  This is nice because it doesn't require the Python SDK to care (and people have been doing this in class methods for
+  factories for a while now). We can even provide a helper that does basically this like so:
+
+    ```python
+    activities = workflow.proxy_activity_class(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
+* We can make it possible to pass `None` as the `self` parameter like this:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, None, "foo")
+    ```
+  And then we could check for that being self and ignore it. However, due to Mypy not yet supporting all of `ParamSpec`,
+  that requires us to have a two-parameter overload version of `execute_activity` _and_ to make the first param optional
+  like so:
+  
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[LocalParamType, LocalParamType2], LocalReturnType],
+        arg: Optional[LocalParamType],
+        arg2: LocalParamType2,
+        /,
+        *,
+    ```
+  This means if you had a real two-param activity, you'd now be able to have an optional first parameter. Also, I am not
+  yet confident how this will look when we remove all of these overloads once Mypy gets their `ParamSpec` support.
+* We could have a completely separate call like so:
+
+    ```python
+    workflow.execute_activity_method(ActivityClass.some_activity, "foo")
+    ```
+* Instead of having `@activity.defn` set an attribute on the function (which is then transitively transferred for others
+  using `functools.wraps`), we could have it return something like `ActivityDefinition[ParamTypes, ReturnType]` (which
+  does not get the transitive benefits). The reason we might want this is then we can have `@activity.method_defn` or
+  similar that is smart enough to convert to an activity definition. I have been trying to avoid actually wrapping
+  functions so far.
+* We could support the following with an overload:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, "foo", method=True)
+    ```
+  This would work with a special overload like:
+
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[Any, LocalParamType], LocalReturnType],
+        arg: LocalParamType,
+        /,
+        *,
+        method: Literal[True],
+    ```
+  This seems like a fairly sane way, but requires callers pass a special parameter to call
```

Comment:
Sane, but, 100% forgettable and if I understand correctly, won't blow up until they actually try to execute the activity if they do

Reply:
It will blow up with the type checker (because it'd pick the other overload which would require two parameters, one for self and one for the real parameter). But we have to remember that many Python developers don't include type checking. There is no common Python compilation step to catch this stuff anyways. Of course at runtime I can do all sorts of things to fix or error in bad situations.